Centralize truth

annevk opened the pull request:
See whatwg/meta#57 and #75.

whatwg/whatwg.org#303 is the corresponding change to whatwg.org to generate documents from the new JSON database.

domenic commented on db.json:
```diff
@@ -0,0 +1,215 @@
+      "standards": [
+        {
+          "title": "Compatibility",
+          "ref-title": "COMPAT",
```
Probably `ref_title` per https://w3ctag.github.io/design-principles/#casing-rules

domenic commented on db.json:
```diff
@@ -0,0 +1,215 @@
+          "ref-title": "COMPAT",
+          "href": "https://compat.spec.whatwg.org/",
+          "description": "The Compatibility Standard describes a collection of non-standard (and often vendor-prefixed) CSS properties and DOM APIs that web browsers need to support for compatibility with the de facto web.",
+          "authors": []
```
Is the idea that authors inherits from workstream editors automatically, or did you just not fill this out yet? The cases like HTML are the tricky ones.

annevk replied:
Yeah, I was thinking authors would add to editors.

domenic replied:
In theory we could have an "editor" responsible for logistics/final decision calls that is not an "author". Maybe not worth worrying about, though.

domenic replied:
If we do make them additive then I think reflecting that in the field name would be good, e.g. `additional_authors` or `non_editor_authors`.

domenic commented on db.json:
```diff
@@ -0,0 +1,215 @@
+      ]
+    }
+  ],
+  "references": {
```
I'd call this something like `other_documents` and try to make the schema similar to the above, being an array with `ref_title`s inside, and `authors` being objects, not strings.

domenic commented on db.py:
```diff
@@ -0,0 +1,46 @@
+import json
+
+def db_to_workstreams_md(db):
+    output = """# WHATWG Workstreams
```
It's not clear where this document would be deployed. If it's to the website, then HTML seems potentially better as an output.

annevk replied:
This is a sample to generate the existing Workstreams.md resource. One way to go would be that initially we don't really change anything, except for where the data originates and then we change/obsolete some outputs, but also happy to merge those. (This will likely take a while anyway as I don't expect everything to be ready and reviewed by this Friday and after that I'll be out for some time.)

domenic replied:
IMO the existing Workstreams.md resource is best replaced. It has two purposes:

1. The source of truth, which the SG controls, and which our agreements can point to.
2. An externally-readable summary for people wondering what the WHATWG workstreams are.

(1) should be replaced by JSON. (2) should be on the website in HTML, I think, but the overlap with https://spec.whatwg.org/ makes it a bit unclear what the right path would be there.

annevk replied:
That's fair, in that case I suspect we want a separate document for Workstreams that's similar to this Markdown resource as they are primarily an IPR device. Keeping the spec namespace focused on standards only seems preferable to me.

domenic replied:
Yeah, that's probably the easiest route. You could imagine something clever that adds little collapsed workstream annotations or something to the side of spec.whatwg.org, or some other merger, but it'd require too much design thinking probably.

annevk commented:
One thing that would be good to add is sorting of everything that is a list.

annevk commented:
@whatwg/sg I forgot to ping you all, but this is ready for review now. https://github.com/whatwg/whatwg.org/pull/303 is the corresponding change that ensures the new `db.json` is the source of truth for all our "list of things" on various whatwg.org pages. We also want to use it for participate.whatwg.org and potentially for other services (about time to get the Twitter bots back up) in the future.

This will make it easier to add new Workstreams and other things and be somewhat confident everything is updated as needed.

domenic commented on db.json:
```diff
@@ -0,0 +1,496 @@
+        }
+      ],
+      "reference": "DOMPARSING",
+      "obsoletedBy": [
```
Not sure if worth it but we could consider aligning on https://w3ctag.github.io/design-principles/#casing-rules by changing this to `obsoleted_by`. (I imagine we'd have to then change it back to `obsoletedBy` in biblio.json, or update the crawler that ingests biblio.json, so it feels kind of pointless.)

foolip commented on db.py:
```diff
@@ -0,0 +1,108 @@
+    normalized = normalize_internal(input)
+    if normalized != input:
+        sys.stderr.write("Input is not normalized.\n")
+        sys.exit(0)
```
Shouldn't this exit with an error, or CI will just pass with some stuff on stderr?

foolip commented on db.py:
```diff
@@ -0,0 +1,108 @@
+
+def normalize_db(db):
+    i = 0
+    for workstream in db["workstreams"]:
```
This is pretty verbose, and looking at it I wonder if there's something special going on but I think not. How about just `db["workstreams"] = [normalize_workstream(w) for w in db["workstreams"]]` or `db["workstreams"] = list(map(normalize_workstream , db["workstreams"]))`?

foolip commented on db.py:
```diff
@@ -0,0 +1,108 @@
+
+def normalize_workstream(workstream):
+    return {
+      "id": workstream["id"],
```
Idea: `str(workstream["id"])` so that validation would fail if it's not a string. But you could go pretty far with validation, and I think the risk of error is low, so take it or leave it :)

foolip replied:
Let's leave it :)

foolip commented on db.py:
```diff
@@ -0,0 +1,108 @@
+        email = editor["email"]
+    return {
+      "name": editor["name"],
+      "email": email
```
`editor.get("email", None)` here would get the same behavior, and still pretty clear I think. (You could remove the `None` but then it's not as obvious.

annevk commented:
Thanks, I addressed those comments and also added list sorting which caught one thing that wasn't sorted. I could also address the remaining nits, including those from @domenic if desired, but not all seem in scope / worth doing to me so I'd need some additional encouragement.

othermaciej commented:
We discussed this issue in an SG call. We'd like to see what the generated workstreams page will look like.

annevk commented:
Sure thing, https://html5.org/temp/workstreams.html is the (current) output. It would be hosted at `https://whatwg.org/workstreams`.

foolip commented:
@whatwg/sg if two more of us can review this, we can go ahead and merge.

(We decided in https://github.com/whatwg/sg/issues/117 that for editorial changes one approval is enough, and that for things that change policy (which this touches) we'll use the supermajority, so 3 approvals.)

othermaciej commented:
OK, two comments after seeing the generated output.

(1) Can we change the output to be formatted more like Workstreams.md? The new format is needlessly less compact, so only fits 3 Workstream descriptions in the space of 4 in the .md format, the indents seem unnecessary (particularly so the multiple levels of indents) and the lack of distinctive font weight makes it harder to scan for headings.

(2) Where does the intro text on the generated workstreams page come from? I presume not from Workstreams.md any more, since that whole file gets deleted in this PR. If it's hardcoded in the generation script, that's probably not good. If it's a separate file now in the whatwg.org website repo, that would be better, but ideally the SG should have the ability to update that intro text if it ever seems useful to do so. Could we instead keep the source of the intro in this repo?

annevk commented:
@domenic adjusted styling somewhat so https://html5.org/temp/workstreams.html is now shorter. I created `Workstreams Introduction.md` so you can still edit the introduction here. Hopefully all is well now.

dbaron commented:
Looks good to me as well -- thanks for getting this done.
